src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs

```diff
@@ -423,10 +423,12 @@ private async Task<WorkUnitResult> ExecuteBucket(TaskHost taskHost, ItemBucket b
                 {
                     // We need to find the task before logging the task started event so that the using task statement comes before the task started event
                     IDictionary<string, string> taskIdentityParameters = GatherTaskIdentityParameters(bucket.Expander);
-                    TaskRequirements? requirements = _taskExecutionHost.FindTask(taskIdentityParameters);
+                    (TaskRequirements? requirements, TaskFactoryWrapper taskFactoryWrapper) = _taskExecutionHost.FindTask(taskIdentityParameters);
+                    string taskAssemblyName = taskFactoryWrapper?.TaskFactoryLoadedType?.LoadedAssemblyName?.FullName;
```

Comment:
The _name_ of the assembly seems to me less useful than its absolute path. Could we log that instead?

Note that we currently already try to do that but lie about it: https://github.com/dotnet/msbuild/issues/584. Can this be massaged to fix that as well?

Reply:
I will use the absolute path instead of name, and I looked at the "path accuracy" problem. If I understood it correctly, `AssemblyTaskLoader` can load different assembly than expected, so we have to postpone logging of the event after we have instance of the task.

That would mean to extract the logic from ctor of `TaskLoggingContext` and log task started event in `InitializeAndExecuteTask` after invocation of `InitializeForBatch` when task host ask for task instance from loader.
OR
Force task host to get task instance early (and outside of task logging context).
OR
Log mismatch between expected and loaded assembly inside task logging context (only if there is a mismatch).
OR
Some better solution :)

I don't have strong opinion nor experience with this MSBuild part, so please share your preference.

Reply:
Discussed with @rainersigwald offline. We settled on logging extra message if there is a mismatch.

Reply:
Sounds like good candidate for exception (if it's easy) as something that looks like a problem ('mismatch detected') for an expected case can be confusing.

Reply:
I can filter out cases when assembly doesn't have a location at all (usually in-memory assemblies). @rainersigwald

Reply:
Works for me.